dropwizard-auth/src/main/java/io/dropwizard/auth/PolymorphicAuthValueFactoryProvider.java

```diff
@@ -0,0 +1,133 @@
+     * @param injector               injector instance
+     * @param principalClassProvider provider of the principal class
+     */
+    @Inject
+    public PolymorphicAuthValueFactoryProvider(
+        MultivaluedParameterExtractorProvider mpep,
+        ServiceLocator injector,
+        PrincipalClassSetProvider<T> principalClassSetProvider
+    ) {
+        super(mpep, injector, Parameter.Source.UNKNOWN);
+        this.principalClassSet = principalClassSetProvider.clazzSet;
+    }
+
+    /**
+     * Return a factory for the provided parameter. We only expect objects of
+     * the type {@link T} being annotated with {@link Auth} annotation.
+     *
+     * @param parameter parameter that was annotated for being injected
+     * @return the factory if annotated parameter matched type
+     */
+    @Override
+    public AbstractContainerRequestValueFactory<?> createValueFactory(Parameter parameter) {
+        if (!parameter.isAnnotationPresent(Auth.class)
+           || !principalClassSet.contains(parameter.getRawType())) {
+            return null;
+        }
+
+        return new AbstractContainerRequestValueFactory<Principal>() {
+
+            /**
+             * @return {@link Principal} stored on the request, or
+             *         {@code null} if no object was found.
+             */
+            public Principal provide() {
+                final Principal principal = getContainerRequest()
+                    .getSecurityContext()
+                    .getUserPrincipal();
+
+                if (principal == null) {
+                    throw new IllegalStateException("Cannot inject a custom principal into unauthenticated request");
+                }
+                return principal;
+            }
+        };
+    }
+
+    @Singleton
+    static class AuthInjectionResolver extends ParamInjectionResolver<Auth> {
+
+        /**
+         * Create new {@link Auth} annotation injection resolver.
+         */
+        public AuthInjectionResolver() {
+            super(PolymorphicAuthValueFactoryProvider.class);
+        }
+    }
+
+    @Singleton
+    protected static class PrincipalClassSetProvider<T extends Principal> {
+
+        private final Set<Class<? extends T>> clazzSet;
+
+        public PrincipalClassSetProvider(Set<Class<? extends T>> clazzSet) {
+            this.clazzSet = clazzSet;
+        }
+    }
+
+    /**
+     * Injection binder for {@link PolymorphicAuthValueFactoryProvider} and
+     * {@link AuthInjectionResolver}.
+     *
+     * @param <T> the type of the principal
+     */
+    public static class Binder<T extends Principal> extends AbstractBinder {
+
+        private final Set<Class<? extends T>> principalClassSet;
+
+        public Binder(Set<Class<? extends T>> principalClassSet) {
+            this.principalClassSet = principalClassSet;
+        }
+
+        @Override
+        protected void configure() {
+            bind(new PrincipalClassSetProvider<>(principalClassSet))
```

Comment:
Nit: I think, it would be better to be consistent with `AuthValueFactoryProvider` and  have each bind statement (apart from the last) on one line.